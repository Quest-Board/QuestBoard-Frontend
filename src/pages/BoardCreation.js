import React, { Component } from "react";
import {Link, Redirect} from "react-router-dom";
import { ReactMultiEmail, isEmail } from 'react-multi-email';
import 'react-multi-email/style.css';

//import icons
import AddIcon from '@material-ui/icons/Add'
import RemoveIcon from '@material-ui/icons/Remove'

export default class BoardCreation extends Component {

    boardName;
    invites;
    boardID;

    constructor(props){
        super(props);
        this.state = {
            redirect:null,
            emails: [],
            columns: ['']
        };
    }

    CreationCall= async (event)=>{
        event.preventDefault();
        //board name
        await fetch('http://coms-319-t15.cs.iastate.edu/api/board/create?boardname=' + this.boardName, {
            method: 'POST',
          })
          .then((response) => {
            if (!response.ok){
                alert("Error creating board")
                return
            }
          }).then(
              (result) => {
                 this.boardID = result.ID; 
              }
          )

        //add column 1
        fetch('http://coms-319-t15.cs.iastate.edu/api/board/addcolumn', {
        method: 'POST',
        body: JSON.stringify({
            category: this.column1Name,
            boardID: this.boardID
            })
        })
        .then((response) => {
        if (!response.ok){
            alert("Error creating column 1, " + this.column1Name)
            return
        }
        })

        //add column 2
        fetch('http://coms-319-t15.cs.iastate.edu/api/board/addcolumn', {
        method: 'POST',
        body: JSON.stringify({
            category: this.column2Name,
            boardID: this.boardID
            })
        })
        .then((response) => {
        if (!response.ok){
            alert("Error creating column 2, " + this.column2Name)
            return
        }
        })

        //add column 3
        fetch('http://coms-319-t15.cs.iastate.edu/api/board/addcolumn', {
        method: 'POST',
        body: JSON.stringify({
            category: this.column3Name,
            boardID: this.boardID
            })
        })
        .then((response) => {
        if (!response.ok){
            alert("Error creating column 3, " + this.column3Name)
            return
        }
        })

        //invites
        for(const email of this.state.emails){
            fetch('http://coms-319-t15.cs.iastate.edu/api/board/addmember/' + this.boardID, {
            method: 'POST',
            body: JSON.stringify({
                email: email
                })
            })
            .then((response) => {
            if (!response.ok){
                alert("Error adding member " + email)
                return
            }
            })
        }
        this.setState({redirect: "/User/Board"})
    }

    BoardNameChangeHandler=(event)=>{
        this.boardName=event.target.value;
    }

    createColumnEntry(){
        return this.state.columns.map((el, i) => 
            <div key={i}>
                <label>Column {i + 1}</label>
                <div style={{display: "block"}}>
                    <input type="text" className="form-control" value={el||''} placeholder="Name" 
                        onChange={this.handleChange.bind(this, i)}
                        style={{width: "70%", display: "inline"}}/>
                    <button className="btn" value='remove' onClick={this.removeClick.bind(this, i)}><RemoveIcon /></button>
                    <button className="btn" value='remove' onClick={this.addClick.bind(this)}><AddIcon /></button>
                </div>
            </div>          
        )
    }

    handleChange(i, event) {
        let columns = [...this.state.columns];
        columns[i] = event.target.value;
        this.setState({ columns });
     }
     
     addClick(){
       this.setState(prevState => ({ columns: [...prevState.columns, '']}))
     }
     
     removeClick(i){
        if(this.state.columns.length > 1){
            let columns = [...this.state.columns];
            columns.splice(i,1);
            this.setState({ columns });
        }
     }

    render() {
        if(this.state.redirect){
            return <Redirect to={this.state.redirect}/>
        }
        const { emails } = this.state;

        return (
            <div className="Board">
                <nav className="navbar navbar-expand-lg navbar-light fixed-top">
                    <div className="container">
                        <Link className="navbar-brand" to={"/User/Board"}>QuestBoard</Link>
                        <div className="collapse navbar-collapse" id="navbarToggler">
                            <ul className="navbar-nav ml-auto">
                                <li className="nav-item">
                                    <Link className="nav-link" to={"/User/Board"}>My Boards</Link>
                                </li>
                                <li className="nav-item">
                                    <Link className="nav-link" to={"/User/CreateBoard"}>Create Board</Link>
                                </li>
                                <li className="nav-item">
                                    <Link className="nav-link" to={"/User/JoinBoard"}>Join Board</Link>
                                </li>
                            </ul>
                        </div>
                    </div>
                </nav>
                <div className="menu-wrapper">
                    <div className="menu">
                        <form onSubmit={this.CreationCall}>
                            <h3>Create Board</h3>
                            <div className="form-group">
                                <label>Board Name</label>
                                <input type="text" className="form-control" placeholder="Name" onChange={this.BoardNameChangeHandler}/>

                                {this.createColumnEntry()}

                                <label>Invite others</label>

                                <ReactMultiEmail
                                    placeholder="email"
                                    emails={emails}
                                    onChange={(_emails) => {
                                        this.setState({ emails: _emails });
                                    }}
                                    validateEmail={email => {
                                        return isEmail(email); // return boolean
                                    }}
                                    getLabel={(
                                        email,
                                        index,
                                        removeEmail,
                                    ) => {
                                        return (
                                        <div data-tag key={index}>
                                            {email}
                                            <span data-tag-handle onClick={() => removeEmail(index)}>
                                            Ã—
                                            </span>
                                        </div>
                                        );
                                    }}
                                />

                            </div>
                            <button type="submit" className="btn btn-primary btn-block" style={{width: '100%'}}>Submit</button>
                        </form>
                    </div>
                </div>
            </div>
        );
    }
}