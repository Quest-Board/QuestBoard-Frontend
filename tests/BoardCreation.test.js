import 'regenerator-runtime/runtime';
import React from 'react';
import renderer from 'react-test-renderer';
import BoardCreation from '../src/pages/BoardCreation';
import { BrowserRouter } from 'react-router-dom';


test('test3', () => {
    const component = renderer.create(
        <BrowserRouter>
            <BoardCreation/>
        </BrowserRouter>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    //manually trigger callback
    //tree.props.onMouseEnter();
    //re-render
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    //manually trigger the callback


});