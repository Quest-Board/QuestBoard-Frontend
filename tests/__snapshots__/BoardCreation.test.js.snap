// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test3 1`] = `
<div
  className="Board"
>
  <nav
    className="navbar navbar-expand-lg navbar-light fixed-top"
  >
    <div
      className="container"
    >
      <a
        className="navbar-brand"
        href="/User/Board"
        onClick={[Function]}
      >
        QuestBoard
      </a>
      <div
        className="collapse navbar-collapse"
        id="navbarToggler"
      >
        <ul
          className="navbar-nav ml-auto"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/User/Board"
              onClick={[Function]}
            >
              My Boards
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/User/CreateBoard"
              onClick={[Function]}
            >
              Create Board
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/User/JoinBoard"
              onClick={[Function]}
            >
              Join Board
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    className="menu-wrapper"
  >
    <div
      className="menu"
      style={
        Object {
          "overflowY": "auto",
        }
      }
    >
      <form
        onSubmit={[Function]}
      >
        <h3>
          Create Board
        </h3>
        <div
          className="form-group"
        >
          <label>
            Board Name
          </label>
          <input
            className="form-control"
            id="boardName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
          />
          <label
            style={
              Object {
                "paddingLeft": 0,
                "paddingTop": 15,
              }
            }
          >
            Board Columns
          </label>
          <button
            className="btn"
            onClick={[Function]}
            type="button"
            value="add"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </button>
          <div
            style={
              Object {
                "paddingLeft": 30,
              }
            }
          >
            <label>
              Column 
              1
            </label>
            <div
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <input
                className="form-control"
                onChange={[Function]}
                placeholder="Name"
                style={
                  Object {
                    "display": "inline",
                    "width": "90%",
                  }
                }
                type="text"
                value=""
              />
              <button
                className="btn"
                onClick={[Function]}
                type="button"
                value="remove"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13H5v-2h14v2z"
                  />
                </svg>
              </button>
            </div>
          </div>
          <label>
            Invite others
          </label>
          <div
            className=" react-multi-email  empty"
            onClick={[Function]}
          >
            <span
              data-placeholder={true}
            >
              email
            </span>
            <input
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          className="btn btn-primary btn-block"
          style={
            Object {
              "width": "100%",
            }
          }
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`test3 2`] = `
<div
  className="Board"
>
  <nav
    className="navbar navbar-expand-lg navbar-light fixed-top"
  >
    <div
      className="container"
    >
      <a
        className="navbar-brand"
        href="/User/Board"
        onClick={[Function]}
      >
        QuestBoard
      </a>
      <div
        className="collapse navbar-collapse"
        id="navbarToggler"
      >
        <ul
          className="navbar-nav ml-auto"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/User/Board"
              onClick={[Function]}
            >
              My Boards
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/User/CreateBoard"
              onClick={[Function]}
            >
              Create Board
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/User/JoinBoard"
              onClick={[Function]}
            >
              Join Board
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    className="menu-wrapper"
  >
    <div
      className="menu"
      style={
        Object {
          "overflowY": "auto",
        }
      }
    >
      <form
        onSubmit={[Function]}
      >
        <h3>
          Create Board
        </h3>
        <div
          className="form-group"
        >
          <label>
            Board Name
          </label>
          <input
            className="form-control"
            id="boardName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
          />
          <label
            style={
              Object {
                "paddingLeft": 0,
                "paddingTop": 15,
              }
            }
          >
            Board Columns
          </label>
          <button
            className="btn"
            onClick={[Function]}
            type="button"
            value="add"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </button>
          <div
            style={
              Object {
                "paddingLeft": 30,
              }
            }
          >
            <label>
              Column 
              1
            </label>
            <div
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <input
                className="form-control"
                onChange={[Function]}
                placeholder="Name"
                style={
                  Object {
                    "display": "inline",
                    "width": "90%",
                  }
                }
                type="text"
                value=""
              />
              <button
                className="btn"
                onClick={[Function]}
                type="button"
                value="remove"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13H5v-2h14v2z"
                  />
                </svg>
              </button>
            </div>
          </div>
          <label>
            Invite others
          </label>
          <div
            className=" react-multi-email  empty"
            onClick={[Function]}
          >
            <span
              data-placeholder={true}
            >
              email
            </span>
            <input
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          className="btn btn-primary btn-block"
          style={
            Object {
              "width": "100%",
            }
          }
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
